File: src/binaryen_stubs_exports.c
#define CAML_NAME_SPACE
#include <caml/mlvalues.h>
#include <caml/fail.h>
#include <caml/memory.h>
#include <caml/alloc.h>

#include "binaryen-c.h"
#include "ocaml_helpers.h"


/* Allocating an OCaml custom block to hold the given BinaryenExportRef */
static value alloc_BinaryenExportRef(BinaryenExportRef exp)
{
  value v = caml_alloc_custom(&binaryen_ops, sizeof(BinaryenExportRef), 0, 1);
  BinaryenExportRef_val(v) = exp;
  return v;
}

CAMLprim value
caml_binaryen_add_function_export(value _module, value _internalName, value _externalName) {
  CAMLparam3(_module, _internalName, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* internalName = Safe_String_val(_internalName);
  char* externalName = Safe_String_val(_externalName);
  BinaryenExportRef exp = BinaryenAddFunctionExport(module, internalName, externalName);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_add_table_export(value _module, value _internalName, value _externalName) {
  CAMLparam3(_module, _internalName, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* internalName = Safe_String_val(_internalName);
  char* externalName = Safe_String_val(_externalName);
  BinaryenExportRef exp = BinaryenAddTableExport(module, internalName, externalName);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_add_memory_export(value _module, value _internalName, value _externalName) {
  CAMLparam3(_module, _internalName, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* internalName = Safe_String_val(_internalName);
  char* externalName = Safe_String_val(_externalName);
  BinaryenExportRef exp = BinaryenAddMemoryExport(module, internalName, externalName);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_add_global_export(value _module, value _internalName, value _externalName) {
  CAMLparam3(_module, _internalName, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* internalName = Safe_String_val(_internalName);
  char* externalName = Safe_String_val(_externalName);
  BinaryenExportRef exp = BinaryenAddGlobalExport(module, internalName, externalName);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_get_export(value _module, value _externalName) {
  CAMLparam2(_module, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* externalName = Safe_String_val(_externalName);
  BinaryenExportRef exp = BinaryenGetExport(module, externalName);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_remove_export(value _module, value _externalName) {
  CAMLparam2(_module, _externalName);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  char* externalName = Safe_String_val(_externalName);
  BinaryenRemoveExport(module, externalName);
  CAMLreturn(Val_unit);
}

CAMLprim value
caml_binaryen_get_num_exports(value _module) {
  CAMLparam1(_module);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  BinaryenIndex num = BinaryenGetNumExports(module);
  CAMLreturn(Val_int(num));
}

CAMLprim value
caml_binaryen_get_export_by_index(value _module, value _index) {
  CAMLparam2(_module, _index);
  BinaryenModuleRef module = BinaryenModuleRef_val(_module);
  BinaryenIndex index = Int_val(_index);
  BinaryenExportRef exp = BinaryenGetExportByIndex(module, index);
  CAMLreturn(alloc_BinaryenExportRef(exp));
}

CAMLprim value
caml_binaryen_export_get_name(value _exp) {
  CAMLparam1(_exp);
  BinaryenExportRef exp = BinaryenExportRef_val(_exp);
  CAMLreturn(caml_copy_string(BinaryenExportGetName(exp)));
}

CAMLprim value
caml_binaryen_export_get_value(value _exp) {
  CAMLparam1(_exp);
  BinaryenExportRef exp = BinaryenExportRef_val(_exp);
  CAMLreturn(caml_copy_string(BinaryenExportGetValue(exp)));
}

CAMLprim value
caml_binaryen_external_function(value _unit) {
  CAMLparam1(_unit);
  CAMLreturn(Val_int(BinaryenExternalFunction()));
}

CAMLprim value
caml_binaryen_external_table(value _unit) {
  CAMLparam1(_unit);
  CAMLreturn(Val_int(BinaryenExternalTable()));
}

CAMLprim value
caml_binaryen_external_memory(value _unit) {
  CAMLparam1(_unit);
  CAMLreturn(Val_int(BinaryenExternalMemory()));
}

CAMLprim value
caml_binaryen_external_global(value _unit) {
  CAMLparam1(_unit);
  CAMLreturn(Val_int(BinaryenExternalGlobal()));
}

CAMLprim value
caml_binaryen_external_tag(value _unit) {
  CAMLparam1(_unit);
  CAMLreturn(Val_int(BinaryenExternalTag()));
}

CAMLprim value
caml_binaryen_export_get_kind(value _exp) {
  CAMLparam1(_exp);
  BinaryenExportRef exp = BinaryenExportRef_val(_exp);
  CAMLreturn(Val_int(BinaryenExportGetKind(exp)));
}

File: src/binaryen_stubs_exports.js
//Provides: caml_binaryen_add_function_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_function_export(wasm_mod, internal_name, external_name) {
  return wasm_mod.addFunctionExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_table_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_table_export(wasm_mod, internal_name, external_name) {
  return wasm_mod.addTableExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_memory_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_memory_export(wasm_mod, internal_name, external_name) {
  return wasm_mod.addMemoryExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_global_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_global_export(wasm_mod, internal_name, external_name) {
  return wasm_mod.addGlobalExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_get_export
//Requires: caml_jsstring_of_string
function caml_binaryen_get_export(wasm_mod, external_name) {
  return wasm_mod.getExport(caml_jsstring_of_string(external_name));
}

//Provides: caml_binaryen_remove_export
//Requires: caml_jsstring_of_string
function caml_binaryen_remove_export(wasm_mod, external_name) {
  return wasm_mod.removeExport(caml_jsstring_of_string(external_name));
}

//Provides: caml_binaryen_get_num_exports
function caml_binaryen_get_num_exports(wasm_mod) {
  return wasm_mod.getNumExports();
}

//Provides: caml_binaryen_get_export_by_index
function caml_binaryen_get_export_by_index(wasm_mod, index) {
  return wasm_mod.getExportByIndex(index);
}

//Provides: caml_binaryen_export_get_name
//Requires: binaryen
//Requires: caml_string_of_jsstring
function caml_binaryen_export_get_name(exp) {
  var export_info = binaryen.getExportInfo(exp);
  return caml_string_of_jsstring(export_info.name);
}

//Provides: caml_binaryen_export_get_value
//Requires: binaryen
//Requires: caml_string_of_jsstring
function caml_binaryen_export_get_value(exp) {
  var export_info = binaryen.getExportInfo(exp);
  return caml_string_of_jsstring(export_info.value);
}

//Provides: caml_binaryen_external_function
//Requires: binaryen
function caml_binaryen_external_function() {
  return binaryen._BinaryenExternalFunction();
}

//Provides: caml_binaryen_external_table
//Requires: binaryen
function caml_binaryen_external_table() {
  return binaryen._BinaryenExternalTable();
}

//Provides: caml_binaryen_external_memory
//Requires: binaryen
function caml_binaryen_external_memory() {
  return binaryen._BinaryenExternalMemory();
}

//Provides: caml_binaryen_external_global
//Requires: binaryen
function caml_binaryen_external_global() {
  return binaryen._BinaryenExternalGlobal();
}

//Provides: caml_binaryen_external_tag
//Requires: binaryen
function caml_binaryen_external_tag() {
  return binaryen._BinaryenExternalTag();
}

//Provides: caml_binaryen_export_get_kind
//Requires: binaryen
function caml_binaryen_export_get_kind(exp) {
  var export_info = binaryen.getExportInfo(exp);
  return export_info.kind;
}
